### PR TITLE
Add ASM Multicluster demo

### PR DESCRIPTION
**Pull request by askmeegs**
Demonstrate how to split the Bank of Anthos services across 2 clusters, and use ASM endpoint discovery to route between the 2 clusters.

### CODE REVIEW

File: extras/asm-multicluster/README.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+kubectl apply -f ../../istio-manifests/frontend-ingress.yaml
+
+kubectx cluster-2 
+kubectl apply -f services-all.yaml 
+kubectl apply -f ../../kubernetes-manifests/config.yaml
+kubectl apply -f ../../extras/jwt/jwt-secret.yaml
+kubectl apply -f ../../kubernetes-manifests/ledger-db.yaml
+kubectl apply -f ../../kubernetes-manifests/balance-reader.yaml
+kubectl apply -f ../../kubernetes-manifests/ledger-writer.yaml
+kubectl apply -f ../../kubernetes-manifests/transaction-history.yaml
```
Comment (Shabirmean) [inline extras/asm-multicluster/README.md:79]: Same comment as above

File: extras/asm-multicluster/README.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+```
+kubectx cluster-1; kubectl get svc istio-ingressgateway -n istio-system 
+
+NAME                   TYPE           CLUSTER-IP     EXTERNAL-IP     PORT(S)                                                                      AGE
+istio-ingressgateway   LoadBalancer   10.7.254.103   <EXTERNAL_IP>  15021:30429/TCP,80:32101/TCP,443:31503/TCP,15012:30573/TCP,15443:31131/TCP   21m
+```
+
+Navigate to the `EXTERNAL_IP` in a browser; you should see the Bank of Anthos login page. 
+
+7. **Open the Google Cloud Console, and navigate to Anthos > Servic Mesh**. View the Bank of Anthos services in the Anthos Service Mesh dashboard. In the Table view, you should see metrics populated for services in both cluster-1 and cluster 2.
```
Comment (Shabirmean) [inline extras/asm-multicluster/README.md:126]: At this point I had to `Enable` Anthos when I navigated to the page in the console. Should we mention it here?

image

**Review by Shabirmean — COMMENTED** [review]
I did a first pass and left some comments. 

Unfortunately when I log in there seems to be some errors when I try to make a deposit. Ill do another pass on it.

**Comment by Shabirmean** [comment]
I retried again and I see only 1 container per pod in cluster-1 when I do 
`kubectx cluster-1; kubectl get pods`

File: extras/asm-multicluster/cluster-setup.sh [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+export CLUSTER_2_NAME="cluster-2"
+export CLUSTER_2_ZONE="us-central1-b"
+############################################################
+
+
+# Note - workload identity is required for ASM 
+echo "☸️ Creating clusters..."
+
+gcloud config set projec ${PROJECT_ID}
+gcloud services enable container.googleapis.com
```
[the remark this answers is not in the document]
Comment (askmeegs) [reply extras/asm-multicluster/cluster-setup.sh:38]: added release channel flag + image type.

File: extras/asm-multicluster/README.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+kubectx cluster-1 
+kubectl apply -f services-all.yaml 
+kubectl apply -f ../../kubernetes-manifests/config.yaml
+kubectl apply -f ../../extras/jwt/jwt-secret.yaml
+kubectl apply -f ../../kubernetes-manifests/accounts-db.yaml 
+kubectl apply -f ../../kubernetes-manifests/contacts.yaml
+kubectl apply -f ../../kubernetes-manifests/frontend.yaml
+kubectl apply -f ../../kubernetes-manifests/userservice.yaml
+kubectl apply -f ../../kubernetes-manifests/loadgenerator.yaml
+kubectl apply -f ../../istio-manifests/frontend-ingress.yaml
```
[the remark this answers is not in the document]
Comment (askmeegs) [reply extras/asm-multicluster/README.md:70]: These instructions assume the user is already in the extras/asm-multicluster directory -- I will note this at the top

File: extras/asm-multicluster/README.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+## Steps 
+
+1. **Set variables.**
+
+```
+export PROJECT_ID="<your-project-id>" 
+```
+
+2. **Run the cluster setup script**. This script creates 2 GKE clusters, `cluster-1` and `cluster-2`, installs Anthos Service Mesh, and sets up cross-cluster endpoint discovery. This script takes about 10 minutes to run. 
+
```
[the remark this answers is not in the document]
Comment (askmeegs) [reply extras/asm-multicluster/README.md:36]: fixed

**Comment by Shabirmean** [comment]
I've tested this a couple of times following steps in the README, however there seems to be no access to _cluster2_ pods from _cluster1_. As in, when I log into the `frontend` it cannot access _transaction_ details which are in a pod from _cluster2_.

**Comment by NickScottKortical** [comment]
Just keeping an eye on this as i think its an example of what i want to replicate, just to check should it have been merged? (it has the do not merge tag still)

**Comment by askmeegs** [comment]
> Just keeping an eye on this as i think its an example of what i want to replicate, just to check should it have been merged? (it has the do not merge tag still)

Hi Nick, we're still doing some debugging on our end, currently the ASM Endpoint Discovery piece is not working with this setup. We'll hopefully get that resolved + merge this demo in soon.

**Comment by Shabirmean** [comment]
The [TLS issue](https://github.com/istio/istio/issues/32523) reported in the istio repository could be related to what we observe here.

File: extras/asm-multicluster/cluster-setup.sh [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+
+echo "🌏 Setting up Endpoint Discovery between clusters..."
+# https://cloud.google.com/service-mesh/docs/managed-control-plane#configure_endpoint_discovery_only_for_multi-cluster_installations 
+
+export CTX_1="cluster-1"
+export CTX_2="cluster-2"
+
+echo "Letting cluster 1 know about cluster 2..."  
+./asm-${CLUSTER_1_NAME}/$REVISION/bin/istioctl x create-remote-secret --context=${CTX_2} --name=cluster2 | kubectl apply -f - --context=${CTX_1}
```
Comment (Shabirmean) [inline extras/asm-multicluster/cluster-setup.sh:100]: I think the names for the secrets here must match the actual cluster names as per the notes [here](https://istio.io/latest/docs/ops/diagnostic-tools/multicluster/#step-by-step-diagnosis)
```sh
--name=${CLUSTER_2_NAME} # for the secret in the first cluster
--name=${CLUSTER_1_NAME} # for the secret in the second cluster
```

Comment (askmeegs) [reply extras/asm-multicluster/cluster-setup.sh:100]: fixed!

**Review by Shabirmean — APPROVED** [review]
LGTM! 🎉 

**Minor:** There is an empty file `extras/asm-multicluster/pwd` which I believe can be removed

**Pull request merged**